### PR TITLE
Measure invalidation performance and fixes

### LINKED ISSUE #24551: Improve measure invalidation and legacy Layout(s) performance
### TL;DR

- Do not bubble up `MeasureInvalidated` on scrollable views and pages
- Do not synchronously measure/arrange children on legacy layout upon child measure invalidated: simply wait for the next native layout pass

### Description of Change

I executed a [speedscope.zip](https://github.com/user-attachments/files/16814038/speedscope.zip) on [davidortinau/AllTheLists](https://github.com/davidortinau/AllTheLists/blob/net9/src/Pages/LearningPage.xaml#L20) `LearningPage` using the latest `8.0.90-ci.net8.24430.1` build.

What I noticed is that #23052 `OnChildMeasureInvalidatedInternal` was showing up and taking 2% of total time (recursive calls).
image

This revealed an interesting thing, the time consumed in there is only due to legacy layouts (which includes `ContentView`) which are triggering synchronous native measurements instead of waiting for the layout pass which is optimized as it runs only once for each invalidated view.

image

Therefore I realized al that code is actually not needed and substantially wrong.

After these changes, the speedscope looks like this:
image

As you can see, legacy layout calls and page calls are gone, except for the `TemplatedCell` of the collection view which is still triggering synchronous measurements.

image

### PR DESCRIPTION
**Pull request by albyrock87**
### Description of Change

Initial proposal to solve #24551 .

I executed a speedscope on [davidortinau/AllTheLists](https://github.com/davidortinau/AllTheLists/blob/net9/src/Pages/LearningPage.xaml#L20) `LearningPage` using the latest nightly `8.0.99-ci.net8.24468.1` build.

What I noticed is that #23052 `OnChildMeasureInvalidatedInternal` was showing up and taking 2% of total time (recursive calls).

What I've done:
- Avoids triggering measure invalidated completely while building the MAUI tree in memory (`Handler = null`)
- Fixes some iOS `SetNeedsLayout` propagation issues and cleans up detection of auto `SetNeedsLayout` parent propagation with `IPropagatesSetNeedsLayout` internal interface
- Removes "out-of-layout-pass" measurements from legacy layouts making them fast as new layouts

### WinUI Speedscope

Kindly provided by @MartyIX

**Before `main`**
image
[1726819823.MAIN.speedscope.json](https://github.com/user-attachments/files/17073463/1726819823.MAIN.speedscope.json)

**After `PR`**
image
[1726819948.PR.speedscope.json](https://github.com/user-attachments/files/17073469/1726819948.PR.speedscope.json)

**Before `main`**
image
[1726819865.MAIN.speedscope.json](https://github.com/user-attachments/files/17073484/1726819865.MAIN.speedscope.json)

**After `PR`**
image
[1726819973.PR.speedscope.json](https://github.com/user-attachments/files/17073488/1726819973.PR.speedscope.json)

### iOS Speedscope

**Before `main`**
image
[before.speedscope.json.zip](https://github.com/user-attachments/files/17048820/before.speedscope.json.zip)

**After `PR`**
image
[after.speedscope.json.zip](https://github.com/user-attachments/files/17049097/after.speedscope.json.zip)

### Issues Fixed

Fixes  #24551

### CODE REVIEW

**Comment by bcaceiro** [comment]
@albyrock87 I'm quite interested on your phrase: " due to legacy layouts (which includes ContentView)" , the example has a CollectionView with a DataTemplate , that has a ContentView as a parent - should we have better performance, in having for instance, DataTemplates with a Grid / Border as a parent?

**Comment by albyrock87** [comment]
@bcaceiro after this PR it shouldn't matter if you use ContentView or not.
Before this PR I can say there is a performance advantage in completely avoid views based on legacy Layout class.

File: src/Controls/src/Core/LegacyLayouts/Layout.cs [hunk 1]
```diff
@@ -192,8 +192,12 @@ private protected override IList<Element> LogicalChildrenInternalBackingStore
 		public override SizeRequest Measure(double widthConstraint, double heightConstraint, MeasureFlags flags = MeasureFlags.None)
 		{
 			SizeRequest size = base.Measure(widthConstraint - Padding.HorizontalThickness, heightConstraint - Padding.VerticalThickness, flags);
-			return new SizeRequest(new Size(size.Request.Width + Padding.HorizontalThickness, size.Request.Height + Padding.VerticalThickness),
-				new Size(size.Minimum.Width + Padding.HorizontalThickness, size.Minimum.Height + Padding.VerticalThickness));
+			var request = new Size(size.Request.Width + Padding.HorizontalThickness, size.Request.Height + Padding.VerticalThickness);
+			var minimum = new Size(size.Minimum.Width + Padding.HorizontalThickness, size.Minimum.Height + Padding.VerticalThickness);
+
+			DesiredSize = request;
```
Comment (albyrock87) [inline src/Controls/src/Core/LegacyLayouts/Layout.cs:198]: Base method sets `DesiredSize` so the fact this was missing was simply wrong.

File: src/Controls/src/Core/Page/Page.cs [hunk 2]
```diff
@@ -583,29 +582,6 @@ protected void UpdateChildrenLayout()
...
-				var logicalChildren = this.InternalChildren;
-				for (var i = 0; i < logicalChildren.Count; i++)
-				{
-					var v = logicalChildren[i] as VisualElement;
-					if (v != null && v.IsVisible && (!v.IsPlatformEnabled || !v.IsPlatformStateConsistent))
-						return;
-				}
-			}
-
-			InvalidateMeasureInternal(InvalidationTrigger.MeasureChanged);
```
Comment (albyrock87) [inline src/Controls/src/Core/Page/Page.cs:606]: This was already not happening due to the fact there was no `MeasureInvalidated` propagation prior #23052

Comment (PureWeen) [reply src/Controls/src/Core/Page/Page.cs:606]: No :-/ I've been wanting to get rid of this code

The main edge case I could see is if someone was using shimmed pages on iOS but in theory those should still follow the maui layout system via the shim wrapper

File: src/Controls/src/Core/VisualElement/VisualElement.cs [hunk 3]
```diff
@@ -1354,10 +1374,38 @@ public void InvalidateMeasureNonVirtual(InvalidationTrigger trigger)
...
 			InvalidateMeasureInternal(trigger);
 		}
+		
+		internal void InvokeMeasureInvalidated(InvalidationTrigger trigger)
+		{
+			MeasureInvalidated?.Invoke(this, new InvalidationEventArgs(trigger));
+		}
 
+		bool _isApplyingBindings;
+		InvalidationTrigger? _applyingBindingsInvalidationTrigger;
```
[the remark this answers is not in the document]
Comment (albyrock87) [reply src/Controls/src/Core/VisualElement/VisualElement.cs]: It seems like the additional memory usage is worth it from my performance tests.
I had to revert this part per @PureWeen request, but I'll think about your feedback on the follow-up PR.
We can always change `InvalidationTrigger` to a `enum InvalidationTrigger: byte`.. it's a public API enum, but is used only on a `IVisualElementController` public method.

**Review by jonathanpeppers — DISMISSED** [review]
I don't really know the details of how MAUI's layout works, but this seems ok to me.

The fields I was asking about are removed now. 👍

**Review by PureWeen — DISMISSED** [review]

File: src/Controls/src/Core/LegacyLayouts/Layout.cs [hunk 4]
```diff
@@ -517,20 +527,13 @@ internal virtual void OnChildMeasureInvalidated(VisualElement child, Invalidatio
 				{
 					return;
 				}
-				if (trigger == InvalidationTrigger.HorizontalOptionsChanged || trigger == InvalidationTrigger.VerticalOptionsChanged)
+				if (trigger is InvalidationTrigger.HorizontalOptionsChanged or InvalidationTrigger.VerticalOptionsChanged)
 				{
 					ComputeConstraintForView(view);
```
Comment (albyrock87) [inline src/Controls/src/Core/LegacyLayouts/Layout.cs:532]: I just realized something..
`LegacyLayout` has this code which invalidates view legacy constraints with `ComputeConstraintForView`
But now, I'm not propagating MeasureInvalidated on *Options changes.
Which means this code will never run, and potentially generates wrong behavior.

So either:
- This entire method can be completely removed (if nowadays `ComputeContraintForView` is some kind of legacy XF code)
- My choice to not propagate `*Options` invalidation is wrong

File: src/Controls/src/Core/LegacyLayouts/Layout.cs [hunk 5]
```diff
@@ -503,7 +513,7 @@ internal virtual void OnChildMeasureInvalidated(VisualElement child, Invalidatio
 			int count = children.Count;
 			for (var index = 0; index < count; index++)
 			{
-				if (LogicalChildrenInternal[index] is VisualElement v && v.IsVisible && (!v.IsPlatformEnabled || !v.IsPlatformStateConsistent))
+				if (LogicalChildrenInternal[index] is VisualElement { IsVisible: true } v && (!v.IsPlatformEnabled || !v.IsPlatformStateConsistent))
```
Comment (albyrock87) [inline src/Controls/src/Core/LegacyLayouts/Layout.cs:516]: This line can very likely be removed thanks to the fact that `IsPlatformEnabled` is now being checked on `VisualElement` before triggering measure invalidation.

**Comment by albyrock87** [comment]
I'd like to wait for #24762 and rebase so we can actually measure performance everywhere :)

**Comment by albyrock87** [comment]
Closing in favor of https://github.com/dotnet/maui/pull/25664

**Pull request closed without merging**